packages/components/src/tools-panel/test/index.js

```diff
@@ -238,6 +238,59 @@ describe( 'ToolsPanel', () => {
 		} );
 	} );
 
+	describe( 'application of is-empty class to panel', () => {
+		it( 'should include is-empty class when no controls displayed', () => {
+			const { container } = render(
+				<ToolsPanel { ...defaultProps }>
+					<ToolsPanelItem { ...altControlProps }>
+						<div>Only optional controls</div>
+					</ToolsPanelItem>
+				</ToolsPanel>
+			);
+
+			const panel = getPanel( container );
+			const isEmpty = panel.classList.contains( 'is-empty' );
```

Comment:
It would be better to [avoid testing implementation details](https://kentcdodds.com/blog/testing-implementation-details).

We'd want to avoid using hardcoded classnames or checking for the existence of the placeholder and, instead, write tests from the point of view of the user as much as possible. For example, in this case, showing/hiding `ToolsPanelItems` and making sure they appear in the same order.

Reply:
Thanks. I've updated the approach in this PR to avoid the hardcoded `is-empty` class. As such, I've also replaced these tests with a single test to assert the order of the visible controls remains consistent.

The remaining tests that involve implementation details are already in place but will be removed in the follow-up to this PR as discussed.